### PR TITLE
Circuit: Update note_commit to take into account asset

### PR DESCRIPTION
**Pull request by ConstanceBeguier**
In the circuit, we update note_commit to take into account asset.
Previously, note_commit returns cm = hash(Q_ZEC, msg) + [rcm]R.
Now, note_commit returns
- cm = hash(Q_ZEC, msg) + [rcm]R for ZEC note
- cm = hash(Q_ZSA, msg || asset) + [rcm]R for ZSA note

We now evaluate note_commit with the following steps
1. evaluate **hash_zec = hash(Q_ZEC, msg)**
2. evaluate **hash_zsa = hash(Q_ZSA, msg || asset)**
3. select **hash = hash_zec if is_native_asset**
                             **= hash_zsa otherwise**
4. evaluate **cm = hash + [rcm]R**
5. check some constraints on msg and asset and their decompositions
6. return **cm**

The following modifications are required to update note_commit:
-  add a is_native_asset witness (and check that it is a boolean and its value is correct according to asset)
- add a MUX chip to evaluate a multiplexer on Pallas points

Warning: we increased the size of the Orchard circuit !

### CODE REVIEW

File: src/constants/sinsemilla.rs [hunk 1]
```diff
@@ -37,6 +37,18 @@ pub const Q_NOTE_COMMITMENT_M_GENERATOR: ([u8; 32], [u8; 32]) = (
...
+    [
+        207, 235, 191, 45, 66, 225, 8, 126, 199, 188, 39, 26, 115, 106, 18, 2, 191, 173, 75, 9, 65,
+        225, 175, 193, 224, 202, 228, 177, 3, 75, 228, 1,
+    ],
+    [
+        220, 251, 80, 86, 182, 182, 99, 67, 254, 97, 241, 22, 79, 111, 161, 176, 79, 97, 208, 98,
+        116, 57, 110, 196, 25, 73, 239, 31, 196, 97, 19, 30,
+    ],
+);
+
```
Comment (PaulLaux) [inline src/constants/sinsemilla.rs:51]: How does this got generated?

Comment (ConstanceBeguier) [reply src/constants/sinsemilla.rs:51]: The test `q_note_zsa_commitment_m` checks that the values are correct.

**Comment by ConstanceBeguier** [comment]
I just rebased on zsa1 branch

File: src/circuit/note_commit.rs [hunk 2]
```diff
@@ -894,6 +1055,9 @@ impl PkdCanonicity {
 
     #[allow(clippy::too_many_arguments)]
     fn assign(
+        // The variable names are for `pk_d`
+        // This function is also used with `asset`.
+        // We have just to replace `pk_d`, `b_3`, `c`, `d_0` by `asset`, `h_2`, `i`, `j_0`
```
[the remark this answers is not in the document]
Comment (PaulLaux) [reply src/circuit/note_commit.rs:1060]: please provide the full context:
instead `This function is also used with asset.` 
use `This function is used for both A and B`
`For A: specifics`, 
`For B: specifics`.

File: src/circuit/note_commit.rs [hunk 3]
```diff
@@ -1622,10 +1792,23 @@ pub(in crate::circuit) mod gadgets {
         let (g, g_0, g_1) =
             DecomposeG::decompose(&lookup_config, chip.clone(), &mut layouter, &rho, &psi)?;
 
-        // h = h_0 || h_1 || h_2
+        // h_zec = h_0 || h_1 || 0000
```
Comment (PaulLaux) [inline src/circuit/note_commit.rs:1795]: why constant `0000`?

File: src/circuit/note_commit.rs [hunk 4]
```diff
@@ -2230,129 +2485,196 @@ mod tests {
                     Value::known(rcm),
                 )?;
 
+                let asset = NonIdentityPoint::new(
+                    ecc_chip.clone(),
+                    layouter.namespace(|| "witness asset"),
+                    self.asset.map(|asset| asset.cv_base().to_affine()),
+                )?;
+
+                let is_native_asset = assign_free_advice(
```
Comment (PaulLaux) [inline src/circuit/note_commit.rs:2494]: This part is repeated in `circuit.rs`, consider encapsulation.

File: src/circuit/note_commit.rs [hunk 5]
```diff
@@ -2230,129 +2485,196 @@ mod tests {
...
+        }
+
         let two_pow_254 = pallas::Base::from_u128(1 << 127).square();
+        let mut rng = OsRng;
+        let random_asset = {
+            let sk = SpendingKey::random(&mut rng);
+            let isk = IssuanceAuthorizingKey::from(&sk);
+            let ik = IssuanceValidatingKey::from(&isk);
+            let asset_descr = "zsa_asset";
+            AssetBase::derive(&ik, asset_descr)
```
Comment (PaulLaux) [inline src/circuit/note_commit.rs:2565]: Repeated here and in `value_commit_orchard.rs`

**Review by PaulLaux — COMMENTED** [review]
added some minor comments. As discussed, let's wait for today's meeting before the final approve.

**Comment by ConstanceBeguier** [comment]
I just rebased on zsa1

File: src/circuit.rs [hunk 6]
```diff
@@ -220,6 +223,9 @@ impl plonk::Circuit<pallas::Base> for Circuit {
         // Constrain v_old = 0 or enable_spends = 1      (https://p.z.cash/ZKS:action-enable-spend).
         // Constrain v_new = 0 or enable_outputs = 1     (https://p.z.cash/ZKS:action-enable-output).
         // Constrain split_flag = 1 or nf_old = nf_old_pub
+        // Constrain is_native_asset to be boolean
+        // Constraint if is_native_asset = 1 then asset = native_asset
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply src/circuit.rs:227]: Done

File: src/circuit.rs [hunk 7]
```diff
@@ -265,7 +286,29 @@ impl plonk::Circuit<pallas::Base> for Circuit {
...
+                    (
+                        "(is_native_asset = 1) =>  (asset_x = native_asset_x)",
+                        is_native_asset.clone() * diff_asset_x.clone(),
+                    ),
+                    (
+                        "(is_native_asset = 1) => (asset_y = native_asset_y)",
+                        is_native_asset.clone() * diff_asset_y.clone(),
+                    ),
+                    // To prove that `asset` is not equal to `native_asset`, we will prove that either
+                    // `x(asset) - x(native_asset)` or `y(asset) - y(native_asset)` is not equal to zero.
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply src/circuit.rs:304]: Yes, I updated the comment

File: src/circuit.rs [hunk 8]
```diff
@@ -795,6 +865,76 @@ impl plonk::Circuit<pallas::Base> for Circuit {
...
+                    .inner()
+                    .y()
+                    .copy_advice(|| "asset_y", &mut region, config.advices[3], 1)?;
+
+                // `diff_asset_x_inv` and `diff_asset_y_inv` will be used to prove that either
+                // `x(asset) - x(native_asset)` or `y(asset) - y(native_asset)` is not equal to zero.
+                // If `(x(asset) - x(native_asset)) * diff_asset_x_inv = 1`, then  `x(asset) - x(native_asset)`
+                // is invertible. Thus, `x(asset) - x(native_asset)` is not equal to zero.
+                // If either `x(asset) - x(native_asset)` or `y(asset) - y(native_asset)` is not
+                // equal to zero, then asset is not equal to native_asset.
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply src/circuit.rs:888]: I simplified this comment.

File: src/circuit/note_commit.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply src/circuit/note_commit.rs:1061]: Done

File: src/circuit/note_commit.rs (hunk 3 above)
Comment (ConstanceBeguier) [reply src/circuit/note_commit.rs]: Updated by creating h_2_zec and h_2_zsa

File: src/circuit/note_commit.rs (hunk 4 above)
Comment (ConstanceBeguier) [reply src/circuit/note_commit.rs]: Created assign_is_native_asset

File: src/circuit/note_commit.rs (hunk 5 above)
Comment (ConstanceBeguier) [reply src/circuit/note_commit.rs]: Created AssetBase::random(...)

**Pull request merged**